# Experimentation

## Data Ingestion

In [1]:
import os
import pandas as pd

In [2]:
pwd

'/Users/macbook/Documents/workspace_datahackerman/tesco_customer_churn'

In [3]:
FILE_NAME = "CustomerChurn.csv"
DATA_FOLDER = "data"

In [4]:
main_path = os.getcwd()
main_path

'/Users/macbook/Documents/workspace_datahackerman/tesco_customer_churn'

In [5]:
file_path = os.path.join(os.path.join(main_path, DATA_FOLDER), FILE_NAME)
file_path

'/Users/macbook/Documents/workspace_datahackerman/tesco_customer_churn/data/CustomerChurn.csv'

In [6]:
%%time

churn_data = pd.read_csv(file_path)

CPU times: user 43 ms, sys: 11.8 ms, total: 54.8 ms
Wall time: 66.9 ms


In [7]:
churn_data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

Adjust column headers to lowercase

In [100]:
churn_data.columns = map(str.lower, churn_data.columns)

In [101]:
churn_data.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [102]:
churn_data.shape

(7043, 21)

## Data Inspection

In [105]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   object 
 4   dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   object 


Only tenure , monthlycharges and totalcharges are Numerical Variables.

churndata.totalcharges data type needs modification

to unify decimal places with Monthly and modify to float data type

In [107]:
churn_data["totalcharges"] = churn_data["totalcharges"].astype(float)

In [108]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   object 
 4   dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   object 


In [109]:
churn_data.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [110]:
churn_data.tail()

customerid  gender  seniorcitizen partner dependents  tenure  \
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     phoneservice     multiplelines internetservice onlinesecurity  ...  \
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     deviceprotection techsupport streamingtv streamingmovies        contract  \
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     paperlessbilling              paymentmethod monthlycharges  totalcharges  \
7038              Yes               Mailed check          84.80       1990.50   
7039              Yes    Credit card (automatic)         103.20       7362.90   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40        306.60   
7042              Yes  Bank transfer (automatic)         105.65       6844.50   

      churn  
7038     No  
7039     No  
7040     No  
7041    Yes  
7042     No  

[5 rows x 21 columns]

to check for missing values

In [111]:
churn_data.isna().sum()

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

to check for duplicates

In [113]:
churn_data.duplicated().sum()

0

## Exploratory Data Analysis

. Statistical Analysis
. Plots
. Correlation

To compute basic descriptive statistics

Calculate average tenure

In [114]:
churn_data["tenure"].mean()

32.37114865824223

Calculate average monthly charge

In [115]:
churn_data["monthlycharges"].mean()

64.76169246059918

Calculate average total charge

In [116]:
churn_data["totalcharges"].mean()

2279.7343035638223

Aggregate tenure, Monthly Charges and Total Charges by gender

In [117]:
churn_data[["gender", "tenure"]].groupby("gender").mean()

tenure
gender           
Female  32.244553
Male    32.495359

In [120]:
churn_data[["gender", "monthlycharges"]].groupby("gender").mean()

monthlycharges
gender                
Female       65.204243
Male         64.327482

In [121]:
churn_data[["gender", "totalcharges"]].groupby("gender").mean()

totalcharges
gender              
Female   2279.918062
Male     2279.554008

Create additional columns

In [123]:
churn_data.tenure.min(),churn_data.tenure.max()

(0, 72)

In [124]:
churn_data.monthlycharges.min(),churn_data.monthlycharges.max()

(18.25, 118.75)

In [125]:
churn_data.totalcharges.min(),churn_data.totalcharges.max()

(0.0, 8684.8)

Group customers based on tenure

In [27]:
# churn_data["tenure_group"]=pd.cut(churn_data["tenure"],
#      bins=[-1,25,48,73],labels=['Low','Medium','High'])

In [28]:
# churn_data["MonthlyCharges_group"]=pd.cut(churn_data["tenure"],
#                                  bins=[18,52,85,119],labels=['Low','Medium','High'])

In [29]:
#churn_data["TotalCharges_group"]=pd.cut(churn_data["TotalCharges"],
 #                                 bins=[0,2895,5790,8685],labels=['Low','Medium','High'])

In [126]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   object 
 4   dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   object 


In [127]:
churn_data.sample(5)

customerid  gender  seniorcitizen partner dependents  tenure  \
6805  4323-OHFOW  Female              1     Yes         No      70   
6852  2408-PSJVE    Male              0     Yes        Yes      44   
5177  3672-YITQD    Male              1     Yes         No      36   
4428  5449-FIBXJ    Male              0     Yes        Yes      19   
642   0970-ETWGE    Male              0      No         No       1   

     phoneservice multiplelines internetservice onlinesecurity  ...  \
6805          Yes           Yes     Fiber optic             No  ...   
6852          Yes            No             DSL             No  ...   
5177          Yes           Yes     Fiber optic             No  ...   
4428          Yes           Yes     Fiber optic             No  ...   
642           Yes            No     Fiber optic             No  ...   

     deviceprotection techsupport streamingtv streamingmovies        contract  \
6805               No         Yes         Yes             Yes        One year   
6852              Yes         Yes          No              No  Month-to-month   
5177               No          No          No             Yes  Month-to-month   
4428              Yes         Yes         Yes             Yes  Month-to-month   
642                No          No         Yes             Yes  Month-to-month   

     paperlessbilling              paymentmethod monthlycharges  totalcharges  \
6805              Yes  Bank transfer (automatic)         105.55       7195.35   
6852              Yes               Mailed check          61.90       2924.05   
5177              Yes           Electronic check          84.10       3021.60   
4428               No           Electronic check         103.30       2012.70   
642               Yes           Electronic check          89.55         89.55   

      churn  
6805     No  
6852     No  
5177    Yes  
4428    Yes  
642     Yes  

[5 rows x 21 columns]

In [32]:
# Compute average tenure charge per customer

In [33]:
# def calculate_average_tenure_charge(TotalCharges,tenure):
   # """This function calculates average tenure charge per customer"""

   # result = churn_data["TotalCharges"] /churn_data["tenure"]

 #   return result

In [34]:
# churn_data["AverageTenureCharge"] = calculate_average_tenure_charge

In [35]:
# churn_data.head()

In [36]:
# churn_data.AverageTenureCharge = churn_data.AverageTenureCharge.str.replace(' ','0')

In [37]:
# churn_data["AverageTenureCharge"] = churn_data["AverageTenureCharge"].astype(float)

In [128]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   object 
 4   dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   object 


In [129]:
churn_data.sample(5)

customerid  gender  seniorcitizen partner dependents  tenure  \
2874  8347-GDTMP  Female              0     Yes         No      64   
2643  3550-SAHFP  Female              0      No         No      33   
5079  2507-QZPQS    Male              0      No         No      15   
3794  4280-DLSHD    Male              0     Yes         No       8   
6708  2272-QAGFO  Female              1      No         No      14   

     phoneservice multiplelines internetservice       onlinesecurity  ...  \
2874          Yes            No              No  No internet service  ...   
2643          Yes           Yes     Fiber optic                   No  ...   
5079          Yes           Yes     Fiber optic                   No  ...   
3794          Yes            No             DSL                   No  ...   
6708          Yes           Yes     Fiber optic                   No  ...   

         deviceprotection          techsupport          streamingtv  \
2874  No internet service  No internet service  No internet service   
2643                   No                   No                   No   
5079                   No                  Yes                   No   
3794                  Yes                   No                   No   
6708                   No                   No                   No   

          streamingmovies        contract paperlessbilling  \
2874  No internet service        Two year               No   
2643                  Yes  Month-to-month              Yes   
5079                  Yes  Month-to-month               No   
3794                   No  Month-to-month              Yes   
6708                   No  Month-to-month              Yes   

                paymentmethod monthlycharges  totalcharges  churn  
2874  Credit card (automatic)          19.45       1225.65     No  
2643         Electronic check          83.35       2757.85    Yes  
5079         Electronic check          95.70       1451.10     No  
3794             Mailed check          54.75        445.85     No  
6708         Electronic check          74.30       1096.25    Yes  

[5 rows x 21 columns]

## Model Building

In [132]:
#churn_data.sample()

## Prepare Dataset for Modelling

In [133]:
churn_data.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

## Train_Test_Split

In [134]:
# 0.3, 0.7

In [135]:
from sklearn.model_selection import train_test_split

In [136]:
from autogluon.tabular import TabularDataset, TabularPredictor

In [137]:
# y = mX + c

In [138]:
X = churn_data[["customerid","gender","seniorcitizen",
                "partner","dependents","tenure",
                "phoneservice","multiplelines","internetservice",
                "onlinesecurity","deviceprotection","techsupport",
                "streamingtv","streamingmovies","contract",
                "paperlessbilling","paymentmethod",
                "monthlycharges","totalcharges"]]

In [139]:
type(X)

pandas.core.frame.DataFrame

In [140]:
X.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      multiplelines internetservice onlinesecurity deviceprotection  \
0  No phone service             DSL             No               No   
1                No             DSL            Yes              Yes   
2                No             DSL            Yes               No   
3  No phone service             DSL            Yes              Yes   
4                No     Fiber optic             No               No   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               paymentmethod  monthlycharges  totalcharges  
0           Electronic check           29.85         29.85  
1               Mailed check           56.95       1889.50  
2               Mailed check           53.85        108.15  
3  Bank transfer (automatic)           42.30       1840.75  
4           Electronic check           70.70        151.65

In [141]:
y = churn_data["churn"]

In [142]:
type(y)

pandas.core.series.Series

In [143]:
y.head()

0     No
1     No
2    Yes
3     No
4    Yes
Name: churn, dtype: object

In [144]:
y[0:5]

0     No
1     No
2    Yes
3     No
4    Yes
Name: churn, dtype: object

In [145]:
train_data, test_data = train_test_split(churn_data,test_size = 0.30, random_state =42)

In [146]:
train_data.shape, test_data.shape

((4930, 21), (2113, 21))

In [147]:
train_data.head()

customerid  gender  seniorcitizen partner dependents  tenure  \
1695  4906-ZHGPK    Male              0     Yes        Yes      54   
1095  7439-DKZTW    Male              0      No         No       1   
3889  2592-HODOV    Male              0      No         No      13   
3667  7826-VVKWT  Female              1     Yes        Yes      24   
2902  1658-TJVOA  Female              1      No         No       6   

     phoneservice multiplelines internetservice       onlinesecurity  ...  \
1695          Yes           Yes             DSL                   No  ...   
1095          Yes            No     Fiber optic                   No  ...   
3889          Yes            No              No  No internet service  ...   
3667          Yes            No     Fiber optic                   No  ...   
2902          Yes            No     Fiber optic                  Yes  ...   

         deviceprotection          techsupport          streamingtv  \
1695                   No                  Yes                  Yes   
1095                   No                   No                  Yes   
3889  No internet service  No internet service  No internet service   
3667                  Yes                   No                  Yes   
2902                   No                   No                   No   

          streamingmovies        contract paperlessbilling  \
1695                   No        One year              Yes   
1095                   No  Month-to-month              Yes   
3889  No internet service        One year               No   
3667                  Yes        Two year              Yes   
2902                   No  Month-to-month              Yes   

                paymentmethod monthlycharges  totalcharges  churn  
1695         Electronic check          70.70       3770.00     No  
1095         Electronic check          80.55         80.55     No  
3889  Credit card (automatic)          19.30        259.65     No  
3667         Electronic check          96.55       2263.45     No  
2902         Electronic check          74.10        450.90     No  

[5 rows x 21 columns]

In [148]:
test_data.head()

customerid  gender  seniorcitizen partner dependents  tenure  \
185   1024-GUALD  Female              0     Yes         No       1   
2715  0484-JPBRU    Male              0      No         No      41   
3825  3620-EHIMZ  Female              0     Yes        Yes      52   
1807  6910-HADCM  Female              0      No         No       1   
132   8587-XYZSF    Male              0      No         No      67   

     phoneservice     multiplelines internetservice       onlinesecurity  ...  \
185            No  No phone service             DSL                   No  ...   
2715          Yes               Yes              No  No internet service  ...   
3825          Yes                No              No  No internet service  ...   
1807          Yes                No     Fiber optic                   No  ...   
132           Yes                No             DSL                   No  ...   

         deviceprotection          techsupport          streamingtv  \
185                    No                   No                   No   
2715  No internet service  No internet service  No internet service   
3825  No internet service  No internet service  No internet service   
1807                  Yes                   No                   No   
132                    No                  Yes                   No   

          streamingmovies        contract paperlessbilling  \
185                    No  Month-to-month              Yes   
2715  No internet service  Month-to-month              Yes   
3825  No internet service        Two year               No   
1807                   No  Month-to-month               No   
132                    No        Two year               No   

                  paymentmethod monthlycharges  totalcharges  churn  
185            Electronic check          24.80         24.80    Yes  
2715  Bank transfer (automatic)          25.25        996.45     No  
3825               Mailed check          19.35       1031.70     No  
1807           Electronic check          76.35         76.35    Yes  
132   Bank transfer (automatic)          50.55       3260.10     No  

[5 rows x 21 columns]

In [149]:
label = 'churn'

In [69]:
print("Summary of churn variable:\n", train_data[label].describe())

Summary of churn variable:
 count     4930
unique       2
top         No
freq      3635
Name: churn, dtype: object


## basic model build

In [166]:
%%time

save_path = 'agModels-predictchurn'
predictor = TabularPredictor(label=label, path=save_path).fit(train_data)

Beginning AutoGluon training ...
AutoGluon will save models to "agModels-predictchurn/"
AutoGluon Version:  0.7.0
Python Version:     3.9.16
Operating System:   Darwin
Platform Machine:   x86_64
Platform Version:   Darwin Kernel Version 20.6.0: Fri Dec 16 00:35:00 PST 2022; root:xnu-7195.141.49~1/RELEASE_X86_64
Train Data Rows:    4930
Train Data Columns: 20
Label Column: churn
Preprocessing data ...
AutoGluon infers your prediction problem is: 'binary' (because only two unique label-values observed).
	2 unique label values:  ['No', 'Yes']
	If 'binary' is not the correct problem_type, please manually specify the problem_type parameter during predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Selected class <--> label mapping:  class 1 = Yes, class 0 = No
	Note: For your binary classification, AutoGluon arbitrarily selected which label-value represents positive (Yes) vs negative (No) class.
	To explicitly set the positive_class, either rename

CPU times: user 48 s, sys: 2.75 s, total: 50.7 s
Wall time: 31.9 s


In [167]:
predictor.leaderboard(silent=True)

model  score_val  pred_time_val  fit_time  \
0              CatBoost      0.814       0.008851  5.517798   
1   WeightedEnsemble_L2      0.814       0.011298  6.451184   
2        NeuralNetTorch      0.802       0.023423  6.683447   
3       NeuralNetFastAI      0.792       0.021748  8.822634   
4      RandomForestEntr      0.784       0.148517  1.298468   
5               XGBoost      0.782       0.012325  0.691018   
6      RandomForestGini      0.774       0.076403  2.297627   
7        ExtraTreesGini      0.774       0.131301  1.530490   
8        ExtraTreesEntr      0.768       0.079651  1.439148   
9        KNeighborsUnif      0.758       0.024328  0.015150   
10       KNeighborsDist      0.732       0.023120  0.017284   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.008851           5.517798            1       True   
1                 0.002447           0.933386            2       True   
2                 0.023423           6.683447            1       True   
3                 0.021748           8.822634            1       True   
4                 0.148517           1.298468            1       True   
5                 0.012325           0.691018            1       True   
6                 0.076403           2.297627            1       True   
7                 0.131301           1.530490            1       True   
8                 0.079651           1.439148            1       True   
9                 0.024328           0.015150            1       True   
10                0.023120           0.017284            1       True   

    fit_order  
0           5  
1          11  
2          10  
3           8  
4           4  
5           9  
6           3  
7           6  
8           7  
9           1  
10          2

## Model Evaluation

load test data to make predictions

In [168]:
y_test = test_data[label]

In [169]:
y_test.head()

185     Yes
2715     No
3825     No
1807    Yes
132      No
Name: churn, dtype: object

In [170]:
y_test = test_data[label]  # values to predict

In [171]:
y_test [0:5]

185     Yes
2715     No
3825     No
1807    Yes
132      No
Name: churn, dtype: object

In [172]:
test_data_nolab = test_data.drop(columns=[label])  # delete label column to avoid cheating

In [173]:
test_data_nolab.head()

customerid  gender  seniorcitizen partner dependents  tenure  \
185   1024-GUALD  Female              0     Yes         No       1   
2715  0484-JPBRU    Male              0      No         No      41   
3825  3620-EHIMZ  Female              0     Yes        Yes      52   
1807  6910-HADCM  Female              0      No         No       1   
132   8587-XYZSF    Male              0      No         No      67   

     phoneservice     multiplelines internetservice       onlinesecurity  \
185            No  No phone service             DSL                   No   
2715          Yes               Yes              No  No internet service   
3825          Yes                No              No  No internet service   
1807          Yes                No     Fiber optic                   No   
132           Yes                No             DSL                   No   

             onlinebackup     deviceprotection          techsupport  \
185                    No                   No                   No   
2715  No internet service  No internet service  No internet service   
3825  No internet service  No internet service  No internet service   
1807                   No                  Yes                   No   
132                    No                   No                  Yes   

              streamingtv      streamingmovies        contract  \
185                    No                   No  Month-to-month   
2715  No internet service  No internet service  Month-to-month   
3825  No internet service  No internet service        Two year   
1807                   No                   No  Month-to-month   
132                    No                   No        Two year   

     paperlessbilling              paymentmethod  monthlycharges  totalcharges  
185               Yes           Electronic check           24.80         24.80  
2715              Yes  Bank transfer (automatic)           25.25        996.45  
3825               No               Mailed check           19.35       1031.70  
1807               No           Electronic check           76.35         76.35  
132                No  Bank transfer (automatic)           50.55       3260.10

In [174]:
save_path

'agModels-predictchurn'

In [175]:
predictor

In [176]:
save_model_predictor = TabularPredictor.load(save_path)

In [177]:
save_model_predictor

In [178]:
y_pred = save_model_predictor.predict(test_data_nolab)

In [179]:
y_pred[0:5]

185     Yes
2715     No
3825     No
1807    Yes
132      No
Name: churn, dtype: object

In [180]:
print("Predictions:  \n", y_pred)
perf = predictor.evaluate_predictions(y_true=y_test, y_pred=y_pred, auxiliary_metrics=True)

Evaluation: accuracy on test data: 0.8031235210601041
Evaluations on test data:
{
    "accuracy": 0.8031235210601041,
    "balanced_accuracy": 0.7124592194125784,
    "mcc": 0.4686569946165508,
    "f1": 0.5864811133200795,
    "precision": 0.6828703703703703,
    "recall": 0.5139372822299652
}


Predictions:  
 185     Yes
2715     No
3825     No
1807    Yes
132      No
       ... 
5522    Yes
6377     No
5500     No
2392     No
6705     No
Name: churn, Length: 2113, dtype: object


In [181]:
save_model_predictor.leaderboard(test_data, silent=True)

model  score_test  score_val  pred_time_test  pred_time_val  \
0              CatBoost    0.803124      0.814        0.025866       0.008851   
1   WeightedEnsemble_L2    0.803124      0.814        0.030563       0.011298   
2               XGBoost    0.803124      0.782        0.090804       0.012325   
3        NeuralNetTorch    0.801704      0.802        0.082186       0.023423   
4       NeuralNetFastAI    0.798864      0.792        0.303365       0.021748   
5      RandomForestGini    0.798391      0.774        0.264374       0.076403   
6      RandomForestEntr    0.793658      0.784        0.248574       0.148517   
7        ExtraTreesGini    0.781827      0.774        0.306667       0.131301   
8        ExtraTreesEntr    0.781354      0.768        0.596075       0.079651   
9        KNeighborsUnif    0.773308      0.758        0.031147       0.024328   
10       KNeighborsDist    0.766209      0.732        0.011551       0.023120   

    fit_time  pred_time_test_marginal  pred_time_val_marginal  \
0   5.517798                 0.025866                0.008851   
1   6.451184                 0.004697                0.002447   
2   0.691018                 0.090804                0.012325   
3   6.683447                 0.082186                0.023423   
4   8.822634                 0.303365                0.021748   
5   2.297627                 0.264374                0.076403   
6   1.298468                 0.248574                0.148517   
7   1.530490                 0.306667                0.131301   
8   1.439148                 0.596075                0.079651   
9   0.015150                 0.031147                0.024328   
10  0.017284                 0.011551                0.023120   

    fit_time_marginal  stack_level  can_infer  fit_order  
0            5.517798            1       True          5  
1            0.933386            2       True         11  
2            0.691018            1       True          9  
3            6.683447            1       True         10  
4            8.822634            1       True          8  
5            2.297627            1       True          3  
6            1.298468            1       True          4  
7            1.530490            1       True          6  
8            1.439148            1       True          7  
9            0.015150            1       True          1  
10           0.017284            1       True          2

In [182]:
test_data.head()

customerid  gender  seniorcitizen partner dependents  tenure  \
185   1024-GUALD  Female              0     Yes         No       1   
2715  0484-JPBRU    Male              0      No         No      41   
3825  3620-EHIMZ  Female              0     Yes        Yes      52   
1807  6910-HADCM  Female              0      No         No       1   
132   8587-XYZSF    Male              0      No         No      67   

     phoneservice     multiplelines internetservice       onlinesecurity  ...  \
185            No  No phone service             DSL                   No  ...   
2715          Yes               Yes              No  No internet service  ...   
3825          Yes                No              No  No internet service  ...   
1807          Yes                No     Fiber optic                   No  ...   
132           Yes                No             DSL                   No  ...   

         deviceprotection          techsupport          streamingtv  \
185                    No                   No                   No   
2715  No internet service  No internet service  No internet service   
3825  No internet service  No internet service  No internet service   
1807                  Yes                   No                   No   
132                    No                  Yes                   No   

          streamingmovies        contract paperlessbilling  \
185                    No  Month-to-month              Yes   
2715  No internet service  Month-to-month              Yes   
3825  No internet service        Two year               No   
1807                   No  Month-to-month               No   
132                    No        Two year               No   

                  paymentmethod monthlycharges  totalcharges  churn  
185            Electronic check          24.80         24.80    Yes  
2715  Bank transfer (automatic)          25.25        996.45     No  
3825               Mailed check          19.35       1031.70     No  
1807           Electronic check          76.35         76.35    Yes  
132   Bank transfer (automatic)          50.55       3260.10     No  

[5 rows x 21 columns]

## Improving Predictive Performance

Our basic model build took 31.9 seconds, we can improve our model by setting a shorter time limit of 25 seconds

In [159]:
#time_limit = 25

We can also set an evaluation metric

In [160]:
#metric = 'roc_auc'

In [163]:
#save_path = 'agModels-predictchurnimproved'

.. and also preset our models for best quality fit

In [184]:
%%time

time_limit = 20
save_path = 'AutogluonModels'
metric = 'roc_auc'
predictor = TabularPredictor(label, eval_metric=metric).fit(train_data, time_limit=time_limit, presets='best_quality')

No path specified. Models will be saved in: "AutogluonModels/ag-20230219_203642/"
Presets specified: ['best_quality']
Stack configuration (auto_stack=True): num_stack_levels=0, num_bag_folds=8, num_bag_sets=20
Beginning AutoGluon training ... Time limit = 20s
AutoGluon will save models to "AutogluonModels/ag-20230219_203642/"
AutoGluon Version:  0.7.0
Python Version:     3.9.16
Operating System:   Darwin
Platform Machine:   x86_64
Platform Version:   Darwin Kernel Version 20.6.0: Fri Dec 16 00:35:00 PST 2022; root:xnu-7195.141.49~1/RELEASE_X86_64
Train Data Rows:    4930
Train Data Columns: 20
Label Column: churn
Preprocessing data ...
AutoGluon infers your prediction problem is: 'binary' (because only two unique label-values observed).
	2 unique label values:  ['No', 'Yes']
	If 'binary' is not the correct problem_type, please manually specify the problem_type parameter during predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Selected class

CPU times: user 31.8 s, sys: 1.93 s, total: 33.7 s
Wall time: 24.2 s


In [186]:
predictor.leaderboard(test_data, silent=True)

model  score_test  score_val  pred_time_test  \
0      WeightedEnsemble_L2    0.860705   0.842965        1.153371   
1          CatBoost_BAG_L1    0.860653   0.842803        0.062715   
2  RandomForestGini_BAG_L1    0.837488   0.820652        0.548476   
3  RandomForestEntr_BAG_L1    0.837081   0.820211        0.522152   
4    ExtraTreesGini_BAG_L1    0.831588   0.808307        0.961456   
5    KNeighborsUnif_BAG_L1    0.751180   0.730969        0.008581   
6    KNeighborsDist_BAG_L1    0.742869   0.720020        0.008434   

   pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  \
0       1.004899  18.090163                 0.011594                0.001209   
1       0.119853  13.890118                 0.062715                0.119853   
2       0.400431   1.309332                 0.548476                0.400431   
3       0.462974   1.607918                 0.522152                0.462974   
4       0.485309   2.320530                 0.961456                0.485309   
5       0.013133   0.013325                 0.008581                0.013133   
6       0.020432   0.014443                 0.008434                0.020432   

   fit_time_marginal  stack_level  can_infer  fit_order  
0           1.268352            2       True          7  
1          13.890118            1       True          5  
2           1.309332            1       True          3  
3           1.607918            1       True          4  
4           2.320530            1       True          6  
5           0.013325            1       True          1  
6           0.014443            1       True          2

Model building results show only 7 models can complete iteration within time limit set

In [91]:
#Bringing it all together

In [187]:
test_data["predicted_churn"] = y_pred

In [188]:
test_data.head()

customerid  gender  seniorcitizen partner dependents  tenure  \
185   1024-GUALD  Female              0     Yes         No       1   
2715  0484-JPBRU    Male              0      No         No      41   
3825  3620-EHIMZ  Female              0     Yes        Yes      52   
1807  6910-HADCM  Female              0      No         No       1   
132   8587-XYZSF    Male              0      No         No      67   

     phoneservice     multiplelines internetservice       onlinesecurity  ...  \
185            No  No phone service             DSL                   No  ...   
2715          Yes               Yes              No  No internet service  ...   
3825          Yes                No              No  No internet service  ...   
1807          Yes                No     Fiber optic                   No  ...   
132           Yes                No             DSL                   No  ...   

              techsupport          streamingtv      streamingmovies  \
185                    No                   No                   No   
2715  No internet service  No internet service  No internet service   
3825  No internet service  No internet service  No internet service   
1807                   No                   No                   No   
132                   Yes                   No                   No   

            contract paperlessbilling              paymentmethod  \
185   Month-to-month              Yes           Electronic check   
2715  Month-to-month              Yes  Bank transfer (automatic)   
3825        Two year               No               Mailed check   
1807  Month-to-month               No           Electronic check   
132         Two year               No  Bank transfer (automatic)   

     monthlycharges totalcharges  churn  predicted_churn  
185           24.80        24.80    Yes              Yes  
2715          25.25       996.45     No               No  
3825          19.35      1031.70     No               No  
1807          76.35        76.35    Yes              Yes  
132           50.55      3260.10     No               No  

[5 rows x 22 columns]

In [189]:
test_data.sample(10)

customerid  gender  seniorcitizen partner dependents  tenure  \
5936  9418-RUKPH  Female              0     Yes        Yes      38   
6422  8668-KNZTI    Male              0      No         No      52   
6049  7634-HLQJR  Female              0     Yes        Yes      47   
2623  3130-ICDUP  Female              0      No        Yes       2   
5589  7253-UVNDW  Female              0      No         No      46   
6388  1963-VAUKV  Female              0      No         No       1   
93    6067-NGCEU  Female              0      No         No      65   
3973  3441-CGZJH  Female              0     Yes        Yes      43   
5391  8124-NZVGJ  Female              0      No         No       6   
5720  2325-WINES  Female              0      No         No      32   

     phoneservice     multiplelines internetservice       onlinesecurity  ...  \
5936          Yes                No              No  No internet service  ...   
6422          Yes                No             DSL                  Yes  ...   
6049          Yes                No              No  No internet service  ...   
2623          Yes               Yes     Fiber optic                  Yes  ...   
5589          Yes                No             DSL                   No  ...   
6388          Yes                No              No  No internet service  ...   
93            Yes               Yes     Fiber optic                  Yes  ...   
3973           No  No phone service             DSL                  Yes  ...   
5391           No  No phone service             DSL                   No  ...   
5720          Yes               Yes     Fiber optic                   No  ...   

              techsupport          streamingtv      streamingmovies  \
5936  No internet service  No internet service  No internet service   
6422                   No                   No                   No   
6049  No internet service  No internet service  No internet service   
2623                   No                   No                   No   
5589                  Yes                   No                   No   
6388  No internet service  No internet service  No internet service   
93                     No                  Yes                  Yes   
3973                  Yes                  Yes                  Yes   
5391                  Yes                  Yes                  Yes   
5720                   No                  Yes                  Yes   

            contract paperlessbilling            paymentmethod monthlycharges  \
5936        Two year              Yes             Mailed check          19.95   
6422        One year               No         Electronic check          53.75   
6049        Two year               No  Credit card (automatic)          20.05   
2623  Month-to-month               No  Credit card (automatic)          80.55   
5589        Two year               No  Credit card (automatic)          54.35   
6388  Month-to-month               No             Mailed check          20.40   
93    Month-to-month              Yes  Credit card (automatic)         111.05   
3973        One year              Yes  Credit card (automatic)          60.40   
5391        One year              Yes             Mailed check          49.25   
5720        One year              Yes  Credit card (automatic)         104.05   

     totalcharges  churn  predicted_churn  
5936       756.40     No               No  
6422      2790.65     No               No  
6049      1016.70     No               No  
2623       188.10     No               No  
5589      2460.15    Yes               No  
6388        20.40    Yes               No  
93        7107.00     No               No  
3973      2640.55     No               No  
5391       255.60     No               No  
5720      3416.85     No               No  

[10 rows x 22 columns]